### PR TITLE
Sample metrics, WIP.

### PR DESCRIPTION
**Pull request by mbebenita**
Depends on #207. This adds the `--metrics` option to the `wasm-as` command.

It annotates a few interesting AST nodes, I can go further if anyone is interested in more detailed stats on the binary format.

| Name                             |    Samples |       Size |    % Total |  Sample Size |
|:---------------------------------|-----------:|-----------:|-----------:|-------------:|
| visitConst                       |     203257 |   632.608K |     26.924 |       3.1871 |
| visitGetLocal                    |     291790 |   570.793K |     24.294 |       2.0031 |
| visitSetLocal                    |     108551 |   212.523K |      9.045 |       2.0048 |
| writeLEB128                      |     512783 |   529.536K |     22.538 |       1.0575 |

* `Samples` Number of events.
* `Size` Total size of those events.
* `% Total` Size / Total File Size
* `Sample Size (Avg.)` - Size / Samples

One, interesting things that pop out here is that const's make up 26% of the total file size, that's a lot. I haven't looked at the distribution of locals but I bet there are a lot of duplicates. Perhaps we need constant pools. Perhaps we need Java style `iconst0`, `iconst1` ... `iconstN` opcodes. I can collect more data here.

### CODE REVIEW

**Comment by JSStats** [comment]
There is discussion on optimizing the encoding of the get_local and set_local immediate index by placing them in their own section (just the indexes), and similarly for constants - I understand this is 'layer-1'. Personally I thought the constant pools in the initial prototype looked useful, and perhaps could be re-visited, but this might need to be considered in light of a layer-1 and perhaps can be handled there.

**Comment by mbebenita** [comment]
Here's the full list of i32.constants (BananaBread -> emscripten -> binaryen), if anyone wants to play around. [i32.const.txt](https://github.com/WebAssembly/binaryen/files/139586/i32.const.txt)

**Comment by JSStats** [comment]
A quick look at the i32.const.txt suggests many are int8 and should have been emitted in i8.const operators. Is binaryen using i8.const?

**Comment by mbebenita** [comment]
It is. The average size is `3.1871`, I guess it's either `2` bytes, or `5` bytes.

File: src/wasm-as.cpp [hunk 1]
```diff
@@ -35,6 +36,11 @@ int main(int argc, const char *argv[]) {
                 o->extra["output"] = argument;
                 Colors::disable();
               })
+      .add("--metrics", "", "prints binary metrics",
```
Comment (kripken) [inline src/wasm-as.cpp]: Perhaps we should just make this a pass, and allow wasm-as to run passes?

Comment (kripken) [inline src/wasm-as.cpp]: Or perhaps we should wait with that for a future refactoring of binaryen-shell that can load and store binaries.

**Comment by kripken** [comment]
Is this a temporary thing? If more permanent, it should be tested.

**Comment by mbebenita** [comment]
I marked it as WIP since it's not complete and was just looking for feedback. If this is something we want I can expand on it and wrap everything that writes to the binary file with `enter/leaveMetric`.

File: src/wasm-as.cpp (hunk 1 above)
Comment (mbebenita) [inline src/wasm-as.cpp]: The metrics code needs to be interleaved with the serialization code. I don't see how this could be moved to a pass.

**Comment by mbebenita** [comment]
Bitrotten, will revive it if needed.

**Pull request closed without merging**